src/github.com/matrix-org/dendrite/appservice/workers/transaction_scheduler.go

```diff
@@ -0,0 +1,226 @@
+	// Timeout for sending a single transaction to an application service.
+	transactionTimeout = time.Second * 15
+	// The current transaction ID. Increments after every successful transaction.
+	currentTransactionID = 0
+)
+
+// SetupTransactionWorkers spawns a separate goroutine for each application
+// service. Each of these "workers" handle taking all events intended for their
+// app service, batch them up into a single transaction (up to a max transaction
+// size), then send that off to the AS's /transactions/{txnID} endpoint. It also
+// handles exponentially backing off in case the AS isn't currently available.
+func SetupTransactionWorkers(
+	appserviceDB *storage.Database,
+	workerStates []types.ApplicationServiceWorkerState,
+) error {
+	// Create a worker that handles transmitting events to a single homeserver
+	for _, workerState := range workerStates {
+		// Don't create a worker if this AS doesn't want to receive events
+		if workerState.AppService.URL != "" {
+			go worker(appserviceDB, workerState)
+		}
+	}
+	return nil
+}
+
+// worker is a goroutine that sends any queued events to the application service
+// it is given.
+func worker(db *storage.Database, ws types.ApplicationServiceWorkerState) {
+	ctx := context.Background()
+
+	// Initialize transaction ID counter
+	var err error
+	currentTransactionID, err = db.GetTxnIDWithAppServiceID(ctx, ws.AppService.ID)
+	if err != nil && err != sql.ErrNoRows {
+		log.WithError(err).Fatalf("appservice %s worker unable to get latest transaction ID from DB",
+			ws.AppService.ID)
+		return
+	}
+
+	// Grab the HTTP client for sending requests to app services
+	client := &http.Client{
+		Timeout: transactionTimeout,
+	}
+
+	// Initial check for any leftover events to send from last time
+	eventCount, err := db.CountEventsWithAppServiceID(ctx, ws.AppService.ID)
+	if err != nil {
+		log.WithError(err).Fatalf("appservice %s worker unable to read queued events from DB",
+			ws.AppService.ID)
+		return
+	}
+
+	// Wait if there are no new events to go out
+	if eventCount == 0 {
+		waitForEvents(&ws)
+	}
+
+	// Loop forever and keep waiting for more events to send
+	for {
+		// Set EventsReady to false for some reason (we just sent events?)
+		ws.Cond.L.Lock()
+		ws.EventsReady = false
+		ws.Cond.L.Unlock()
+
+		eventIDs, events, err := db.GetEventsWithAppServiceID(ctx, ws.AppService.ID, transactionBatchSize)
+		if err != nil {
+			log.WithError(err).Errorf("appservice %s worker unable to read queued events from DB",
+				ws.AppService.ID)
+
+			// Wait a little bit for DB to possibly recover
+			time.Sleep(transactionBreakTime)
+			continue
+		}
+
+		// Batch events up into a transaction
+		transactionJSON, err := createTransaction(events)
+		if err != nil {
+			log.WithError(err).Fatalf("appservice %s worker unable to marshal events",
+				ws.AppService.ID)
+
+			return
+		}
+
+		// Send the events off to the application service
+		err = send(client, ws.AppService, transactionJSON)
```

Comment:
As `transactionJSON` is recomputed every time it is possible that a later request might contain more events then the previous one - even with having the same transaction ID.
This might lead to loosing events as the app-service might deduplicate on the first event so will never handle the events which came in between the first and later tries